src/index/compat.rs

```diff
@@ -0,0 +1,186 @@
+use crate::gucs::planning::ENABLE_PGVECTOR_COMPATIBILITY;
+use libc::c_void;
+use pgrx::pg_sys::pfree;
+use pgrx::pg_sys::AsPgCStr;
+use std::collections::HashMap;
+use std::ffi::CStr;
+
+macro_rules! swap_destroy {
+    ($target:expr, $value:expr, $var:ident) => {
+        let $var = $target;
+        $target = $value;
+        pfree($var.cast());
+    };
+}
+
+pub unsafe fn pgvector_stmt_rewrite(pstmt: *mut pgrx::pg_sys::PlannedStmt) {
+    let enabled = ENABLE_PGVECTOR_COMPATIBILITY.get();
+    if !enabled {
+        return;
+    }
+    let utility_statement = (*pstmt).utilityStmt;
+    let is_index = pgrx::is_a(utility_statement, pgrx::pg_sys::NodeTag::T_IndexStmt);
+    if is_index {
+        let istmt = utility_statement.cast();
+        rewrite_type_options(istmt);
+        rewrite_opclass(istmt);
+    }
+}
+
+unsafe fn rewrite_type_options(istmt: *mut pgrx::pg_sys::IndexStmt) {
+    let original = options_from_list((*istmt).options);
+    if let Some(options) = original.get("options") {
+        log::warn!(
+            "Options field '{}' is invalid at compatibility mode.",
+            options
+        );
+    }
+    match CStr::from_ptr((*istmt).accessMethod).to_str() {
+        Ok("hnsw") => {
+            let m = original
+                .get("m")
+                .unwrap_or(&String::from("16"))
+                .parse::<u32>()
+                .unwrap();
+            let ef_construction = original
+                .get("ef_construction")
+                .unwrap_or(&String::from("64"))
+                .parse::<usize>()
+                .unwrap();
+            let arg = pgrx::pg_sys::makeString(
+                format!(
+                    "[indexing.hnsw]\nm = {}\nef_construction = {}",
+                    m, ef_construction
+                )
+                .as_pg_cstr(),
+            );
+            let elem = pgrx::pg_sys::makeDefElem("options".as_pg_cstr(), arg as _, -1);
+            let options = pack_element_list(elem);
+            (*istmt).options = options;
+            swap_destroy!((*istmt).accessMethod, "vectors".as_pg_cstr(), am);
+        }
+        Ok("ivfflat") => {
+            let nlist = original
+                .get("list")
+                .unwrap_or(&String::from("100"))
+                .parse::<u32>()
+                .unwrap();
+            let arg =
+                pgrx::pg_sys::makeString(format!("[indexing.ivf]\nnlist = {}", nlist).as_pg_cstr());
+            let elem = pgrx::pg_sys::makeDefElem("options".as_pg_cstr(), arg as _, -1);
+            let options = pack_element_list(elem);
+            (*istmt).options = options;
+            swap_destroy!((*istmt).accessMethod, "vectors".as_pg_cstr(), am);
+        }
+        Ok(other) => {
+            log::warn!("IndexType '{other}' unrecognized, will not be overwritten");
+        }
+        Err(e) => {
```

Comment:
When could a `access_method` be not UTF-8?

Reply:
https://www.postgresql.org/docs/current/sql-syntax-lexical.html

> Quoted identifiers can contain any character, except the character with code zero. (To include a double quote, write two double quotes.) This allows constructing table or column names that would otherwise not be possible, such as ones containing spaces or ampersands. The length limitation still applies.